### PR TITLE
Add ability to constraint variables to binary or integer values.

### PR DESCRIPTION
**Pull request by stelmo**
Allows construction of MILPs

### CODE REVIEW

File: src/bound.jl [hunk 1]
```diff
@@ -10,8 +10,31 @@ const IntervalBound = Tuple{Float64,Float64}
...
+struct BinaryBound end
+
+const Binary = BinaryBound()
+
+"""
+$(TYPEDEF)
+
+A special type of bound where the variable may only take on integer values.
+"""
+struct IntegerBound end
```
Comment (exaexa) [inline src/bound.jl:25]: does JuMP have any support for limiting the integers to a given interval? or better have that as a separate constraint?

File: src/bound.jl [hunk 2]
```diff
@@ -10,8 +10,31 @@ const IntervalBound = Tuple{Float64,Float64}
...
-Shortcut for possible bounds: either no bound is present (`nothing`), or a
-single number is interpreted as an exact equality bound, or a tuple of 2
-numbers is interpreted as an interval bound.
+A special type of bound where the variable may only take on the values 0 or 1
+exclusively.
 """
-const Bound = Union{Nothing,Float64,IntervalBound}
+struct BinaryBound end
+
+const Binary = BinaryBound()
```
Comment (exaexa) [inline src/bound.jl:18]: not a fan of the name (JuMP naming sucks hard here)

File: src/bound.jl [hunk 3]
```diff
@@ -10,8 +10,31 @@ const IntervalBound = Tuple{Float64,Float64}
...
+
+"""
+$(TYPEDEF)
+
+A special type of bound where the variable may only take on integer values.
+"""
+struct IntegerBound end
+
+const Integers = IntegerBound()
```
Comment (exaexa) [inline src/bound.jl:27]: no docs?

File: src/bound.jl [hunk 4]
```diff
@@ -10,8 +10,31 @@ const IntervalBound = Tuple{Float64,Float64}
...
+"""
+$(TYPEDEF)
+
+Shortcuts for possible bounds:
+
+- either no bound is present (`nothing`),
+- a single number is interpreted as an exact equality bound,
+- a tuple of 2 numbers is interpreted as an interval bound,
+- setting `BinaryBound()` or its alias, `Binary` creates an integer valued variable, that can only take on 0 or 1,
+- setting `IntegerBound()` or its alias, `Integers` creates an integer valued variable.
```
Comment (exaexa) [inline src/bound.jl:38]: `@ref`, also no need to elaborate on what is alias of what here.

File: src/bound.jl (hunk 2 above)
Comment (exaexa) [reply src/bound.jl:18]: (also docs)

File: docs/src/mixed-integer-optimization.jl [hunk 5]
```diff
@@ -0,0 +1,83 @@
+
+# # Example: Mixed integer optimization
+#
+# In this example we demonstrate the use of binary, and integer valued
+# variables. We assume that the reader is already familiar with the construction
+# of `ConstraintTree`s; if not, it is advisable to read the previous part of the
+# documentation first.
+
+# The simple problem we will solve is:
```
Comment (exaexa) [inline docs/src/mixed-integer-optimization.jl:9]: The equation is valid but seems pretty arbitrary, I'd love to have a realistic problem. No worries here, will think of something.

(Ideas: salesman/Hamiltonian? Backpack problem with a few extra constraints?)

**Review by exaexa — COMMENTED** [review]
a few comments, will fix/merge later

File: src/bound.jl (hunk 1 above)
Comment (stelmo) [reply src/bound.jl:25]: I think you just bound then later in constraints

File: src/bound.jl (hunk 2 above)
Comment (stelmo) [reply src/bound.jl:18]: Yep, the naming was tough, especially `Integers`, which is too close to Julia's native `Integer` type I think

File: src/bound.jl (hunk 3 above)
Comment (stelmo) [reply src/bound.jl:27]: Docs are for people who are too lazy to read the source code

File: docs/src/mixed-integer-optimization.jl (hunk 5 above)
Comment (stelmo) [reply docs/src/mixed-integer-optimization.jl:9]: Ah, actually, I went for super simple to just illustrate it and not have too much supporting code which is tiring to read

File: src/bound.jl (hunk 2 above)
Comment (stelmo) [reply src/bound.jl:18]: Other options are `OneOrZero` and `EveryInterger` or something (JuMP has similar things as as MOI sets)

**Comment by exaexa** [comment]
OK so, this is gonna be harder:
- JuMP and other systems don't really seem to support actual integer CONSTRAINTS, just integer VARIABLES. That kinda sucks from the algebraic point of view but what can we do, solver reality is harsh.
- there are very many thinkable types of constraints and their interpretations -- I can think of few representations just for the integers: (bool, choice from vector, choice from range, free integer, powers of 2, approximate free integer with error term on it, .......)

In the long term I suggest we remove the closed type family `Bound` and convert it to open, just as was done for `Value`. That prevents us from forcing a given MILP representation to people, and allows everyone to treat the MILPs as they seem fit. The modification of actual CT source code will thus stay minimal and the functionality will be preserved.

For the current state, I think the easiest way of supporting integer _constraints_ is to just smash in an extra JuMP variable and an equality bound (in the end that would be the same). That should go to the tutorial.

Also, we should re-think the Floatiness of the whole CTs at that point (people might just want to get integers out).

**Comment by stelmo** [comment]
Okay, I think opening it to integer constraints is definitely more general.

**Comment by exaexa** [comment]
@stelmo can you review here?

**Review by stelmo — COMMENTED** [review]
This all looks good to me!

**Pull request merged**